base/compiler/abstractinterpretation.jl

```diff
@@ -755,6 +756,11 @@ function abstract_eval(@nospecialize(e), vtypes::VarTable, sv::InferenceState)
                     t = UnionAll(v, t)
                 end
             end
+            # TODO: the above UnionAll wrapping should be sufficient to prevent this,
+            # but in some cases seems not to be.
```

Comment:
This code is all kinds of wrong. I have a corrected version on my closure-replacement branch moved to a helper function

Reply:
Sure